File: patterns/debounce/rules-python/debounce_generic.py
import threading

from openhab import rule, Registry
from openhab.triggers import ItemStateChangeTrigger

timers = {}


@rule(triggers=[
    ItemStateChangeTrigger("Person1_PresenceSensors"),
    ItemStateChangeTrigger("Person2_PresenceSensors"),
])
class PresenceDetectionDebounce:
    def execute(self, module, input):
        event = input.get("event")
        if not event:
            return

        item_name = event.getItemName()
        triggering_item = Registry.getItem(item_name)

        if item_name in timers:
            timers[item_name].cancel()

        delay = 0 if str(triggering_item.getState()) == "ON" else 2 * 60

        def on_expire(name=item_name, item=triggering_item):
            proxy_name = name.split("_")[0] + "_Present"
            proxy_item = Registry.getItem(proxy_name)
            current_state = str(item.getState())
            if str(proxy_item.getState()) != current_state:
                proxy_item.sendCommand(current_state)
            del timers[name]

        t = threading.Timer(delay, on_expire)
        t.start()
        timers[item_name] = t
        self.logger.info(item_name + ": Debounce-Timer gestartet, Verzoegerung=" + str(delay) + "s")

File: patterns/debounce/rules-js/debounce-generic.js
const timers = new Map();

rules.JSRule({
  name: "Presence detection debounce",
  triggers: [
    triggers.ItemStateChangeTrigger("Person1_PresenceSensors"),
    triggers.ItemStateChangeTrigger("Person2_PresenceSensors")
  ],
  execute: (event) => {
    const triggeringItem = items.getItem(event.itemName);

    if (timers.has(triggeringItem.name)) {
      timers.get(triggeringItem.name).cancel();
    }

    const delay = triggeringItem.state === "ON" ? 0 : 2;
    const t = actions.ScriptExecution.createTimer(
      time.ZonedDateTime.now().plusMinutes(delay),
      () => {
        const proxyName = triggeringItem.name.split("_")[0] + "_Present";
        const proxyItem = items.getItem(proxyName);
        if (proxyItem.state !== triggeringItem.state) {
          proxyItem.sendCommand(triggeringItem.state);
        }
        timers.delete(triggeringItem.name);
      }
    );
    timers.set(triggeringItem.name, t);
  }
});
